drivers/clock_control/clock_control_nrf.c

```diff
@@ -295,12 +295,21 @@ static void generic_hfclk_start(void)
 
 static void generic_hfclk_stop(void)
 {
-	if (atomic_and(&hfclk_users, ~HF_USER_GENERIC) & HF_USER_BT) {
-		/* bt still requesting the clock. */
-		return;
+	/* It's not enough to use only atomic_and() here for synchronization,
+	 * as the thread could be preempted right after that function but
+	 * before hfclk_stop() is called and the preempting code could request
+	 * the HFCLK again. Then, the HFCLK would be stopped inappropriately
+	 * and hfclk_user would be left with an incorrect value.
+	 */
+	unsigned int key = irq_lock();
+
+	hfclk_users &= ~HF_USER_GENERIC;
+	/* Skip stopping if BT is still requesting the clock. */
+	if (!(hfclk_users & HF_USER_BT)) {
+		hfclk_stop();
```

Comment:
I have not tested on target with ISR CPU time profiling but I am concerned with `nrfx_clock_hfclk_stop` that has `while` loop without putting CPU in idle.

Any reason you can think that prevents the clock from being stopped in a asynchronous way?

Reply:
Well, currently there is no other option of stopping the clock using the nrfx driver. I'm not sure about the reasons and if an asynchronous way could be added. Perhaps @nika-nordic or @nordic-krch could say something more in this regard. But anyway, this is actually beyond the scope of this PR.

Reply:
> this is actually beyond the scope of this PR.

Well the change in this PR is enclosing a `while` loop inside `irq_lock()` and `irq_unlock()` which seems a bit dangerous in terms of OS being prevented from covering if anything goes wrong.

Reply:
There are no peripheral events associated with the clock being stopped so asynchronous clock stopping cannot be done in `nrfx_clock`

Reply:
Isn't it that we can stuck in this loop in two case: loop is preempted and someone starts the clock after preemption or hardware failure. Lock prevents first scenario and regarding the second one, we are screwed anyway.